HV-Crash support using panic devices

varunrsekar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
This PR introduces support for Panic devices in libvirt domxml. This enables libvirt to receive panic notification from a QEMU guest. (See https://libvirt.org/formatdomain.html#id112)

Before this PR:
When a windows VMI crashes, we do not get any signal from the guest. So it reports the VM as running even though the guest is crashed.
```
NAME                         AGE   PHASE    IP                        NODENAME   READY
vmi-nocloud-nopanic   12s   Running   10.244.196.152   node01            True

NAME                                                         READY   STATUS    RESTARTS   AGE
virt-launcher-vmi-nocloud-nopanic-v5nzn   3/3          Running    0                   61s

domstate:
sh-5.1# virsh domstate 1
running

guest-info:
sh-5.1# virsh qemu-agent-command 1 '{"execute":"guest-info"}'
error: Guest agent is not responding: QEMU guest agent is not connected
command terminated with exit code 1
```
After this PR:
We can now configure a windows VMI with a hyperv panic device:
```
VMI spec:
...
spec:
  domain:
    devices:
      panicDevices:
      - model: hyperv
...
```
This signals when the guest crashes causing the VMI to go to Failed state.
```
NAME                        AGE     PHASE     IP                        NODENAME   READY
vmi-nocloud-panic     6m17s  Failed       10.244.196.153   node01           False

NAME                                                     READY   STATUS      RESTARTS   AGE
virt-launcher-vmi-nocloud-panic-6k5ns   0/3          Completed   0                  76s

Crash event:
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event Domain event=\"unknown\" detail=\"\" with event id 8 reason 0 received","pos":"client.go:470","timestamp":"2024-09-30T16:30:49.508410Z"}
{"component":"virt-launcher-monitor","level":"info","msg":"Reaped pid 81 with status 0","pos":"virt-launcher-monitor.go:198","timestamp":"2024-09-30T16:30:49.575663Z"}
{"component":"virt-launcher","level":"info","msg":"DomainLifecycle event Domain event=\"stopped\" detail=\"crashed\" with event id 5 reason 2 received","pos":"client.go:470","timestamp":"2024-09-30T16:30:49.709688Z"}
{"component":"virt-launcher","level":"info","msg":"kubevirt domain status: Shutoff(5):Crashed(3)","pos":"client.go:297","timestamp":"2024-09-30T16:30:49.710769Z"}
{"component":"virt-launcher","level":"info","msg":"Domain name event: default_vmi-nocloud-panic","pos":"client.go:424","timestamp":"2024-09-30T16:30:49.712009Z"}


libvirt logs:
2024-09-30 16:30:49.506+0000: panic hyper-v: arg1='0x5a', arg2='0x1', arg3='0x1', arg4='0x0', arg5='0x0'
2024-09-30 16:30:49.507+0000: shutting down, reason=crashed
2024-09-30T16:30:49.524570Z qemu-kvm: terminating on signal 15 from pid 21 (/usr/sbin/virtqemud)

```
<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->

### Why we need it and why it was done in this way
The following tradeoffs were made: N/A

The following alternatives were considered: N/A

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer
<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Feature: Support for defining panic devices in VirtualMachineInstances to catch crash signals from the guest.
```

varunrsekar commented:
/cc rthallisey

lyarwood commented on staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:
```diff
@@ -101,6 +101,12 @@ type VirtualMachineInstancetypeSpec struct {
+	//
+	// +optional
+	// +listType=atomic
+	Panics []v1.Panic `json:"panics,omitempty"`
```
We only use instance types for attributes related to host related resources such as vCPUs, memory, GPUs and HostDevices etc. Everything else that's software defined like vTPMs, EFI etc resides in our preference CRDs. Can you move this over there as a `spec.devices.preferredPanics` attribute?

lyarwood commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -497,6 +497,10 @@ type Devices struct {
+	// Panics provides additional crash information when domain crashes.
+	// +optional
+	// +listtype=atomic
+	Panics []Panic `json:"panics,omitempty"`
```
`panics` isn't a great name IMHO, what about `panicDevices`?

lyarwood commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -609,6 +613,10 @@ type VGPUDisplayOptions struct {
 }
 
+type Panic struct {
+	Model string `json:"model,omitempty"`
```
Can you make this a typed string with some constants defining the supported values? For example:

https://github.com/kubevirt/kubevirt/blob/09a45ff11041456e54f5c2a596c5975ed57e840c/staging/src/kubevirt.io/api/core/v1/schema.go#L30-L38

lyarwood commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -988,6 +988,19 @@ func Convert_v1_Usbredir_To_api_Usbredir(vmi *v1.VirtualMachineInstance, domainD
+
+	for _, panic := range panics {
+		panicDev := api.PanicDevice{
+			Model: panic.Model,
```
`Model` is optional right?

lyarwood reviewed (changes requested):
Thanks for the PR, from an instance type API point of view this attribute should really be in the preference CRD. I'm also not sold on the naming of the attribute in the core API. Would you also mind breaking this up into more targeted commits for easier review?

varunrsekar commented:
> Thanks for the PR, from an instance type API point of view this attribute should really be in the preference CRD. I'm also not sold on the naming of the attribute in the core API. Would you also mind breaking this up into more targeted commits for easier review?

@lyarwood Sure. Thanks for the quick review! I'll reorganize the commits and address the comments by tomorrow.

lyarwood commented on pkg/instancetype/instancetype.go:
```diff
@@ -1386,6 +1387,41 @@ func applyInputPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePre
+		return
+	}
+
+	emptyPanicDevice := virtv1.PanicDevice{}
```
nit - I'd inline this below

lyarwood commented on pkg/instancetype/instancetype.go:
```diff
@@ -1386,6 +1387,41 @@ func applyInputPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePre
+	panicDeviceModels := make(map[string]struct{})
+	for _, panicDevice := range vmiSpec.Domain.Devices.PanicDevices {
+		if panicDevice.Model == nil {
+			foundVMIEmptyPanicDevice = true
```
Could you just use the following to represent an device without a model? Would remove the need for these `bools`:

```
const nilModel = "nilModel"
[..]
panicDeviceModels[nilModel] = struct{}{}
```

varunrsekar replied:
Thanks. I've reworked that code with the hopefully non-conflicting "nilModel" string.

lyarwood commented on pkg/instancetype/instancetype.go:
```diff
@@ -1386,6 +1387,41 @@ func applyInputPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePre
+	if foundPreferredEmptyPanicDevice && !foundVMIEmptyPanicDevice {
+		vmiSpec.Domain.Devices.PanicDevices = append(vmiSpec.Domain.Devices.PanicDevices, emptyPanicDevice)
+	}
+}
```
Would you mind adding some unit tests for this? Appreciate the tests are a mess at the moment but we will be breaking them up during the 1.5 dev cycle finally.

varunrsekar replied:
I've updated the unit-tests. Also added some api validations to the vmi spec. I'm yet to add validations for the preference spec. I'll do that and update here

varunrsekar replied:
@lyarwood I've added api validations for the panic device specs.

lyarwood commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -48,6 +48,14 @@ const (
+const (
+	PanicDeviceModelHyperv  PanicDeviceModel = "hyperv"
+	PanicDeviceModelIsa     PanicDeviceModel = "isa"
+	PanicDeviceModelPvpanic PanicDeviceModel = "pvpanic"
```
Would you mind dropping the repeated type of `PanicDeviceModel` from each of these constants assuming the resulting names don't conflict with anything?

https://libvirt.org/formatdomain.html#panic-device also suggests `s390` as a valid model, we do support `s390x` so I think we should add that here.

varunrsekar replied:
```
This feature is always enabled for:
* pSeries guests, since it's implemented by the guest firmware
* S390 guests, since it's an integral part of the S390 architecture
For the guest types listed above, libvirt automatically adds a panic element to the domain XML.
```
Both pseries and s390 have this feature always enabled by libvirt. We wouldnt need to explicitly add them. Additionally, libvirt doesn't allow you to specify even an address for them:
```
Most users don't need to specify an address, and doing so is forbidden altogether for s390, pseries and hyperv models.
```

varunrsekar replied:
> Would you mind dropping the repeated type of PanicDeviceModel from each of these constants assuming the resulting names don't conflict with anything?
I can't drop the type here. go is inferring the other consts as strings, if i do so.

varunrsekar replied:
ohh got it! I misunderstood the ask earlier.
 I've removed the type prefix from the constants now.

lyarwood commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -609,6 +621,10 @@ type VGPUDisplayOptions struct {
 }
 
+type PanicDevice struct {
+	Model *string `json:"model,omitempty"`
```
```suggestion
	Model *PanicDeviceModel `json:"model,omitempty"`
```

lyarwood replied:
I also assume we don't want to support `address`?

https://libvirt.org/formatdomain.html#panic-device

varunrsekar replied:
If needed, this support can be added later. But I dont see an usecase for it to justify carrying it around. This is only valid for `isa` and `pvpanic` models. And even there they use a default address if none is provided

lyarwood commented:
Apologies just back from PTO, I'll take a deeper look at this tomorrow!

vladikr commented:
/sig compute

vladikr commented:
/test all

vladikr commented:
Thanks for the PR @varunrsekar. Great to see the panic device addition.

varunrsekar commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -48,6 +48,14 @@ const (
+type PanicDeviceModel string
+
+const (
+	Hyperv  PanicDeviceModel = "hyperv"
+	Isa     PanicDeviceModel = "isa"
+	Pvpanic PanicDeviceModel = "pvpanic"
```
> Are we sure that we need to support all of these?

I'm not sure about the need for `isa` and `pvpanic`. I added them since they're supported by libvirt.

> Also, how are we going to treat s390?

s390 doesn't need any support. libvirt automatically adds panic devices for s390 (See https://libvirt.org/formatdomain.html#id112)

vladikr replied:
What happens when a user mistakenly sets the panic device to a value that is not supported by s390, do we need any kind of validation?

varunrsekar replied:
Do you mean validations to not allow panic devices in the VMI spec when its architecture is s390x?

varunrsekar replied:
@vladikr Can you clarify your question? In general, qemu/libvirt will reject unsupported panic devices in the domxml.

xpivarc replied:
We should validate possible values per architecture

varunrsekar replied:
I've added a validation to not allow panic devices for s390 and p-series architectures.

lyarwood reviewed (commented):
Quick nit but something we should follow up with within `pkg/instancetype`.

The unit test failure looks related. I also wonder if we need to provide a simple functional test case here ensuring the VM/VMI moves to `Failed` when the guest panics?

varunrsekar commented on pkg/instancetype/instancetype.go:
```diff
@@ -1478,6 +1479,32 @@ func applyInputPreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePre
 }
 
+func applyPanicDevicePreferences(preferenceSpec *instancetypev1beta1.VirtualMachinePreferenceSpec, vmiSpec *virtv1.VirtualMachineInstanceSpec) {
+	if len(preferenceSpec.Devices.PreferredPanicDevices) == 0 {
```
+1. I wanted to use DevicePreferences in the input. But it didnt fit the pattern in the file. I can do some cleanup in a follow-up

varunrsekar commented:
/retest

varunrsekar commented:
> The unit test failure looks related.

@lyarwood Had to undo the testdata changes and create a separate unittest. Will squash it into commit 5 after review..

> I also wonder if we need to provide a simple functional test case here ensuring the VM/VMI moves to Failed when the guest panics?

Do you know how I can setup that functional test? I don't know how to trigger a guest panic for a windows VM.

lyarwood commented:
/ok-to-test

> > The unit test failure looks related.
> 
> @lyarwood Had to undo the testdata changes and create a separate unittest. Will squash it into commit 5 after review..

ACK thanks.

> > I also wonder if we need to provide a simple functional test case here ensuring the VM/VMI moves to Failed when the guest panics?
> 
> Do you know how I can setup that functional test? I don't know how to trigger a guest panic for a windows VM.

We wouldn't use Windows as the guest OS here, you should be able to trigger the expected behaviour using Fedora with the guest agent installed and sysrq to panic the OS etc.

lyarwood reviewed (commented):
/test all

varunrsekar commented:
> We wouldn't use Windows as the guest OS here, you should be able to trigger the expected behaviour using Fedora with the guest agent installed and sysrq to panic the OS etc.

@lyarwood ACK I was able to locally test a crashing fedora with an ISA panic device:

```
virsh qemu-agent-command 1 '{"execute": "guest-exec", "arguments": {"path": "/bin/sh", "arg": ["-c", "echo c > /proc/sysrq-trigger"], "capture-output": true}}'
```

```
[14:25:46] ./cluster-up/kubectl.sh get vmi
selecting docker as container runtime
NAME         AGE   PHASE    IP               NODENAME   READY
vmi-fedora   40s   Failed   10.244.196.155   node01     False

[14:31:29] ./cluster-up/kubectl.sh get events | grep Stopped
5m53s       Warning   Stopped                 virtualmachineinstance/vmi-fedora    The VirtualMachineInstance crashed.
```
I'll create a functional test in `tests/vmi_lifecycle_test.go` with this.

lyarwood reviewed (changes requested):
Final issue with the use of the guest agent to cause the panic and a supernit that can be a follow up. Otherwise this lgtm.

Can you also add a release note in the [description](https://github.com/kubevirt/kubevirt/pull/13103#issue-2601847343)?

varunrsekar commented on tests/vmi_lifecycle_test.go:
```diff
@@ -485,6 +485,34 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+				defer cancel()
+
+				By("Crashing the vm guest")
+				out := libpod.QemuAgentCommand(vmi, `{"execute": "guest-exec", "arguments": {"path": "/bin/sh", "arg": ["-c", "echo c > /proc/sysrq-trigger"], "capture-output": true}}`)
```
Thanks! Used the console to trigger the panic

varunrsekar commented on tests/vmi_lifecycle_test.go:
```diff
@@ -485,6 +485,34 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+		Context("when guest crashes", decorators.VMIlifecycle, func() {
+			It("should be stopped and have Failed phase when a PanicDevice is provided", func() {
+				vmi := libvmifact.NewFedora()
+				vmi.Spec.Domain.Devices.PanicDevices = []v1.PanicDevice{{Model: pointer.P(v1.Isa)}}
```
Added that helper util in libvmi

xpivarc commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -497,6 +505,10 @@ type Devices struct {
 	// DownwardMetrics creates a virtio serials for exposing the downward metrics to the vmi.
 	// +optional
 	DownwardMetrics *DownwardMetrics `json:"downwardMetrics,omitempty"`
+	// PanicDevices provides additional crash information when a QEMU guest crashes.
```
`PanicDevices provides additional crash information when a guest crashes.`

xpivarc commented on staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:
```diff
@@ -519,6 +519,11 @@ type DevicePreferences struct {
 	// +optional
 	PreferredInterfaceMasquerade *v1.InterfaceMasquerade `json:"preferredInterfaceMasquerade,omitempty"`
+
+	// PreferredPanicDevices optionally defines any panic devices to be used to signal QEMU guest failures.
```
`// PreferredPanicDevices optionally defines any panic devices to be used to signal guest failures.`

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -485,6 +485,37 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 			})
 		})
 
+		Context("when guest crashes", Label("crashing-guest"), decorators.VMIlifecycle, func() {
```
Why the label?

varunrsekar replied:
Sorry this was meant only for my testing. Will remove it

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -485,6 +485,37 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+					&expect.BSnd{S: "sudo su -\n"},
+					&expect.BExp{R: "#"},
+					&expect.BSnd{S: `echo c > /proc/sysrq-trigger` + "\n"},
+					&expect.BExp{R: "#"},
```
Wouldn't this flake in case the guest crashes before a new line is printed?

varunrsekar replied:
The test was failing if I didn't expect a new line. I'm suspecting the console wasn't sending the command without L503.

varunrsekar replied:
```
  ESC[34mSent:ESC[39m "echo c > /proc/sysrq-trigger\n"
  ESC[32mMatch for RE:ESC[39m "echo c > /proc/sysrq-trigger((?s).*)sysrq triggered crash" found: ["echo c > /proc/sysrq-trigger\r\n[   11.766720] sysrq: Trigger a crash\r\n[   11.768184] Kernel panic - not syncing: sysrq triggered crash" "\r\n[   11.766720] sysrq: Trigger a crash\r\n[   11.768184] Kernel panic - not syncing: "] Buffer:  echo c > /proc/sysrq-trigger
  [   11.766720] sysrq: Trigger a crash
  [   11.768184] Kernel panic - not syncing: sysrq triggered crash
```
That `#` was incorrect to look for. I twas just working out because the character was eventually getting printed out in `dmesg` after the crash. 

I've changed it to wait for `sysrq triggered crash` instead. This should help avoid flakes

xpivarc commented on tests/vmi_lifecycle_test.go:
```diff
@@ -485,6 +485,37 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
+				Expect(event.Message).To(ContainSubstring(`The VirtualMachineInstance crashed`), "VMI should be stopped because of a guest crash")
+
+				By("Checking that VirtualMachineInstance has 'Failed' phase")
+				Eventually(func() v1.VirtualMachineInstancePhase {
+					vmi, err := kubevirt.Client().VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred(), "Should get VMI successfully")
+					return vmi.Status.Phase
+				}, 10, 1).Should(Equal(v1.Failed), "VMI should be failed")
```
`Eventually(matcher.ThisVMI(vmi)).WithTimeout(10*time.Second).WithPolling(time.Second).Should(matcher.BeInPhase(v1.Failed))`

xpivarc commented on pkg/virt-launcher/virtwrap/converter/converter_test.go:
```diff
@@ -728,6 +728,12 @@ var _ = Describe("Converter", func() {
+		It("should use panic devices if requested", func() {
+			vmi.Spec.Domain.Devices.PanicDevices = []v1.PanicDevice{{Model: kubevirtpointer.P(v1.Hyperv)}}
+			xml := vmiToDomainXML(vmi, c)
+			Expect(strings.Contains(xml, `<panic model="hyperv"></panic>`)).To(BeTrue(), xml)
```
`Expect(xml).To(ContainSubstring(....`

xpivarc commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -2298,3 +2303,31 @@ func validateCPUHotplug(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpe
 	return causes
 }
+
+func validatePanicDeviceModel(field *k8sfield.Path, model *v1.PanicDeviceModel) *metav1.StatusCause {
```
This would be sufficient for x86.

xpivarc commented on pkg/virt-launcher/virtwrap/converter/testdata/domain_ppc64le.xml.tmpl:
```diff
@@ -122,6 +122,7 @@
       <driver error_policy="stop" name="qemu" type="raw"></driver>
       <alias name="ua-sysprep_secret"></alias>
     </disk>
+    <panic model="hyperv" />
```
Where is this coming from?

varunrsekar replied:
Its not present anymore.

varunrsekar commented:
> pull-kubevirt-e2e-k8s-1.31-sig-performance
This test suite passed. Not sure why the job is marked as FAILED.

> pull-kubevirt-e2e-k8s-1.31-sig-compute-serial
This failed at some setup steps for CRIO

varunrsekar commented:
/retest

rthallisey commented:
/ok-to-test

rthallisey commented:
/retest

varunrsekar commented:
@lyarwood Had to rebase. Can you approve again?

varunrsekar commented:
@vladikr @lyarwood Can I please get some attention on this PR?

vladikr commented:
It looks good, from my point of view.
I think this is a needed feature, and the API looks good to me.
/approve
/hold

I'd like @iholder101 to look at this as well before it goes in.

iholder101 commented on pkg/virt-launcher/virtwrap/converter/converter.go:
```diff
@@ -914,6 +914,19 @@ func Convert_v1_Usbredir_To_api_Usbredir(vmi *v1.VirtualMachineInstance, domainD
 }
 
+func convertPanicDevices(panicDevices []v1.PanicDevice) []api.PanicDevice {
+	domainPanicDevices := make([]api.PanicDevice, 0)
```
nit: a redundant allocation
```suggestion
	var domainPanicDevices []api.PanicDevice
```

iholder101 commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -2335,3 +2340,44 @@ func validateCPUHotplug(field *k8sfield.Path, spec *v1.VirtualMachineInstanceSpe
+		arch = config.GetDefaultArchitecture()
+	}
+
+	if virtconfig.IsS390X(arch) || virtconfig.IsPPC64(arch) {
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: fmt.Sprintf("custom panic devices are not supported on %s architecture", arch),
+			Field:   field.Child("domain", "devices", "panicDevices").String(),
+		})
+	}
```
There's something a bit confusing to me about the fact that these archs would always have a panic device enabled, yet it will always be non-populated at the vmi level. This means that the existence of panic devices cannot be deduced from the vmi spec, perhaps misleading to think that such VMs don't support a panic device.

Should we consider to always populate the panic device field for these archs?

varunrsekar replied:
By including it in the spec, we're indicating the "user"/"kubevirt" is doing something to enable the panic devices. But its the guest OS/libvirt that provides that functionality for those specific architectures. And even the domain XML will not include the panic device. Its just something the S390X architecture/pseries firmware supports inherently.

lyarwood replied:
What about reflecting this in status? That avoids any mutations to the VM on s390x to include a panic device that's going to be added by default by libvirt but still lets the end user see that one has been added to the underlying guest.

varunrsekar replied:
Since they aren't gonna be visible in the domain XML, why would we need to show it?

lyarwood replied:
Are they still visible if we query the domain? If not then ignore me but if they are we could report both the requested and defaults back through the status of the VMI and VM.

iholder101 replied:
@varunrsekar 

> By including it in the spec, we'd be indicating the "user"/"kubevirt" is doing something to enable the panic devices.

A VMI is basically the expanded version of a VM. Usually, users will not be dealing with VMIs directly, but with VM objects. What we do in general in Kubevirt is keep the VM's spec clean but expand and set values in VMI. I think here we can follow the same principle.

> even the domain XML will not include the panic device. Its just something the S390X architecture/pseries firmware supports inherently.

Please correct me if I'm wrong, but IIUC libvirt allows setting this panic device at the domXML, but it would practically serve as a no-op. This means we can set it there, right?

In any case I don't feel very strongly about this. @vladikr WDYT?

varunrsekar replied:
@iholder101 So for s390x/pseries, other models like `isa` would cause compatibility issues and break the VM. So its not really a no-op. Adding a `<panic model="s390" />` would work but it is a no-op. So its not very useful to try to support adding panic devices to such VMs.

I'ld prefer to support only the bare minimum in this feature and extend further as and when usecases arise.

varunrsekar replied:
> Are they still visible if we query the domain? If not then ignore me but if they are we could report both the requested and defaults back through the status of the VMI and VM.
@lyarwood I'm not sure actually. I dont think I've a setup to try to run a s390x VM to check. But since crash detection is an inherent feature of an s390x guest, its not very useful to expose it to the user (even if its exposed in the domxml which I doubt). Thoughts?

iholder101 reviewed (commented):
Generally looks good to me, see comments below.

In addition, should we consider to set a panic device by default? I'm not sure it's possible to do agnostically to the guest's OS.

fossedihelm commented on pkg/libvmi/vmi.go:
```diff
@@ -289,6 +289,12 @@ func WithTPM(persistent bool) Option {
 	}
 }
 
+func WithPanicDevice(model v1.PanicDeviceModel) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Devices.PanicDevices = append(vmi.Spec.Domain.Devices.PanicDevices, v1.PanicDevice{Model: &model})
+	}
+}
```
Please move this to `/libvmi/devices.go` file

fossedihelm reviewed (commented):
Thank you!
Can you put this behind a FG?

varunrsekar commented on staging/src/kubevirt.io/api/core/v1/schema.go:
```diff
@@ -613,6 +625,14 @@ type VGPUDisplayOptions struct {
+	// Model specifies what type of panic device is provided.
+	// The panic model used when this attribute is missing depends on the hypervisor and guest arch.
+	// One of: isa, hyperv, pvpanic.
+	// +optional
```
I already have this validation implemented in the webhook. See [commit](https://github.com/kubevirt/kubevirt/pull/13103/commits/60b41864f8e6672dd6a6162931adaa4b5a22aba7)
I saw [PRs](https://github.com/kubevirt/kubevirt/pull/11156)/[Issues](https://github.com/kubevirt/kubevirt/issues/12998) discussing whether we should use kubebuilder validations instead of the webhook ones. I didnt see a clear consensus on the topic. So I'd rather not make this change until that's sorted. 

This can be in a followup if that's the direction we go though.

fossedihelm commented on tests/vmi_lifecycle_test.go:
```diff
@@ -452,6 +452,10 @@ var _ = Describe("[rfe_id:273][crit:high][vendor:cnv-qe@redhat.com][level:compon
 		})
 
 		Context("when guest crashes", decorators.VMIlifecycle, func() {
+			BeforeEach(func() {
```
This requires the test being `Serial`

varunrsekar replied:
Added the `Serial `tag and verified the test runs successfully.

lyarwood commented on pkg/instancetype/preference/apply/device.go:
```diff
@@ -102,3 +103,30 @@ func applyInputPreferences(preferenceSpec *v1beta1.VirtualMachinePreferenceSpec,
+		panicDeviceModels[model] = struct{}{}
+	}
+
+	for _, preferredPanicDevice := range preferenceSpec.Devices.PreferredPanicDevices {
+		model := nilModel
+		if preferredPanicDevice.Model != nil {
+			model = *preferredPanicDevice.Model
+		}
+		if _, found := panicDeviceModels[model]; !found {
+			vmiSpec.Domain.Devices.PanicDevices = append(vmiSpec.Domain.Devices.PanicDevices, preferredPanicDevice)
+		}
+	}
```
Yeah apologies for missing this, we should really make `PanicDevices` a pointer to a slice to better handle this.

* `nil` - User has provided no input so just fill everything from the preference
* `&[]virt1.PanicDevice{}` - The user explicitly doesn't want any panic devices so ignore the preference
* `&[]virtv1.PanicDevice{virtv1.Hyperv, virtv1.ISA}` - Merge the users requests with the preference 

@fossedihelm @varunrsekar does that make sense?

fossedihelm replied:
> &[]virtv1.PanicDevice{virtv1.Hyperv, virtv1.ISA} - Merge the users requests with the preference

I am not sure about the last one; To me, if users provide a panic device list (even empty), they simply don't care about the preference one.
Why should we merge? In this case, users cannot opt-out of some specific panic device defined in the preference.

lyarwood replied:
@fossedihelm  yeah true that's cleaner, @varunrsekar are you okay with this?

varunrsekar replied:
Sure this makes sense. I'll make the change

varunrsekar replied:
Instead of having it as a pointer to a slice, would it make sense to differentiate the 3 cases directly?
```
// if vmi.Spec.Domain.Devices.PanicDevices == nil, apply preferences
// if len(vmi.Spec.Domain.Devices.PanicDevices) >= 0, don't apply preferences
```

lyarwood replied:
Yes sorry if it's only the two conditions then this doesn't need to be a pointer.

varunrsekar replied:
nvm. Went on an etcd goose chase to figure having explicit empty slices in specs is equivalent to nil slices. So we would need to use a pointer to a slice like you mentioned.
But as an API, it feels kind of weird.
Would it be better to change the preference API to just:
```
PreferredPanicDeviceModel *v1.PanicDeviceModel `json:"preferredPanicDeviceModel,omitempty"`
```
And then apply that preference to any panic devices created with an empty model field?
Example:
```
preference:
  spec:
    devicePreferences:
      preferredPanicDeviceModel: "hyperv"
```
Case 1: No/Empty panic devices: Don't apply any preference
```
vmi:
  spec:
    domain:
      devices:
        panicDevices: []
```

Case 2: Panic Devices without model: Apply preference to both
```
vmi:
  spec:
    domain:
      devices:
        panicDevices:
        - {}
        - model: ""
```
@fossedihelm @lyarwood Thoughts?

lyarwood replied:
That's essentially what we do for network interface models and disk buses so yes if that still provides what you ultimately need I'm okay with this.

varunrsekar replied:
Updated the preference API to take this as input:
```
PreferredPanicDeviceModel *v1.PanicDeviceModel `json:"preferredPanicDeviceModel,omitempty"`
```
If a panic device is specified without a `model` in the vmi spec, then the `preferredPanicDeviceModel` is applied to it.
Eg:
```
apiVersion: instancetype.kubevirt.io/v1beta1
kind: VirtualMachinePreference
metadata:
  name: test-preference
spec:
  devicePreferences:
    preferredPanicDeviceModel: "hyperv"
```
1. Case 1: No panic devices
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vm
spec:
  preference:
    kind: VirtualMachinePreference
    name: test-preference
  template:
    spec:
      domain:
        devices:
          panicDevices: []
```
VMI: no change
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vm-vmi-1
spec:
  domain:
    devices:
      panicDevices: []
```

2. Case 2: 1 panic device with nil model
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vm-vmi-1
spec:
  preference:
    kind: VirtualMachinePreference
    name: test-preference
  template:
    spec:
      domain:
        devices:
          panicDevices: 
          - {}
```
VMI: `hyperv` model added to the panic device
```
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vm
spec:
  domain:
    devices:
      panicDevices:
      - model: hyperv
```

rthallisey commented:
/remove-lifecycle stale

varunrsekar commented on pkg/virt-api/webhooks/validating-webhook/admitters/preference-admitter.go:
```diff
@@ -143,3 +144,26 @@ func admitPreference(request *admissionv1.AdmissionRequest, resource string) *ad
+	var causes []metav1.StatusCause
+	if len(devicePreferences.PreferredPanicDevices) == 0 {
+		return causes
+	}
```
Dropping this since I've switched to a simpler `PreferredPanicDeviceModel` preference api.

varunrsekar commented on pkg/virt-api/webhooks/validating-webhook/admitters/preference-admitter.go:
```diff
@@ -143,3 +144,26 @@ func admitPreference(request *admissionv1.AdmissionRequest, resource string) *ad
+	causes = append(causes, validatePreferredPanicDevices(field.Child("devices"), spec.Devices)...)
+
+	return causes
+}
```
Dropping this since I've switched to a simpler `PreferredPanicDeviceModel` preference api.

varunrsekar commented:
Made a couple changes along with the rebase to ToT:
* Reworked the preference api to only take a `PreferredPanicDeviceModel`: [Panic device support: instancetype v1beta1 API](https://github.com/kubevirt/kubevirt/pull/13103/commits/597b916c2459587b37b6d4defb77e4fd0ec41be4)
* Dropped the preference api validations commit based on the above change
* Added the `Serial` tag to the functest: [func test for panic device](https://github.com/kubevirt/kubevirt/pull/13103/commits/b4b149f50bc0ba325f55443ccd84803179408a09)
* `virtconfig.Is<ARCH>` methods went away so used the arch strings directly in [Panic device support: core v1 api validations](https://github.com/kubevirt/kubevirt/pull/13103/commits/0f40543a17ed1ca4258eed67914a9c282692aabc)

varunrsekar commented:
Rebased to ToT

fossedihelm commented on pkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:
```diff
@@ -2415,6 +2415,15 @@ func validatePanicDevices(field *k8sfield.Path, spec *v1.VirtualMachineInstanceS
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeFieldValueInvalid,
+			Message: "Panic Devices feature gate is not enabled in kubevirt-config",
+			Field:   field.Child("PanicDevices").String(),
```
```suggestion
			Field:   field.Child("domain", "devices", "panicDevices"),
```

fossedihelm commented on pkg/virt-config/featuregate/active.go:
```diff
@@ -95,6 +95,11 @@ const (
 	SecureExecution = "SecureExecution"
+
+	// Owner: @varunrsekar
+	// Alpha: v1.5.0
```
// Alpha: v1.6.0

varunrsekar commented on pkg/instancetype/preference/apply/device_test.go:
```diff
@@ -295,4 +297,40 @@ var _ = Describe("Preference.Devices", func() {
+		vmi.Spec.Domain.Devices.PanicDevices = []virtv1.PanicDevice{isaPanicDeviceTest, emptyPanicDeviceTest}
+		Expect(vmiApplier.ApplyToVMI(field, instancetypeSpec, preferenceSpec, &vmi.Spec, &vmi.ObjectMeta)).To(Succeed())
+		Expect(vmi.Spec.Domain.Devices.PanicDevices[0]).To(Equal(isaPanicDeviceTest))
+		Expect(vmi.Spec.Domain.Devices.PanicDevices[1]).To(Equal(emptyPanicDeviceTest))
```
Removed this redundant test

varunrsekar commented on pkg/instancetype/preference/apply/device_test.go:
```diff
@@ -209,6 +210,7 @@ var _ = Describe("Preference.Devices", func() {
 		Expect(vmi.Spec.Domain.Devices.NetworkInterfaceMultiQueue).
 			To(HaveValue(Equal(*preferenceSpec.Devices.PreferredNetworkInterfaceMultiQueue)))
 		Expect(vmi.Spec.Domain.Devices.BlockMultiQueue).To(HaveValue(Equal(*preferenceSpec.Devices.PreferredBlockMultiQueue)))
+		Expect(vmi.Spec.Domain.Devices.PanicDevices).To(Equal(preferenceSpec.Devices.PreferredPanicDevices))
```
Thanks. Fixed it. 

Somehow I missed running the unittests after the rebase and ended up with a lot of stale code. Thanks for catching it.

varunrsekar commented:
> Thanks @vamsikrishna-siddu! Some comments below.
> Maybe something goes wrong with the rebase and other changes.
> Please, ensure the changes are consistent with `make` and `make generate` and `make test`.
> Would be great having this before the FF. Thanks

Wrong handle btw :sweat_smile: 

* Fixed all the rebasing errors
* Fixed incomplete Preference API changes
* Rebased to ToT

fossedihelm commented:
@varunrsekar Could you please address the lint issue? Thanks

fossedihelm commented:
/test pull-kubevirt-unit-test

varunrsekar commented:
@fossedihelm Fixed the linter error in [apply panic device instancetype preference to vmi spec](https://github.com/kubevirt/kubevirt/pull/13103/commits/e6086078539017b6caa567a701a2182d430778e9).

The other test failures seem unrelated to this PR

fossedihelm commented:
/test pull-kubevirt-build
/test pull-kubevirt-generate
/test pull-kubevirt-code-lint
/test pull-kubevirt-verify-go-mod
/test pull-kubevirt-unit-test
/test pull-kubevirt-manifests
/test pull-kubevirt-check-unassigned-tests

lyarwood commented:
> @lyarwood do you want to take another look?
> leaving the unhold up to you

ACk thanks, apologies for the delay!

/lgtm
/hold cancel

lyarwood commented:
/retest-required

fossedihelm commented:
/retest-required

fossedihelm commented:
/unhold

fossedihelm commented:
/retest-required

fossedihelm commented:
/unhold
